### PR TITLE
LPS-116683 Hide ckeditor toolbar labels if toolbar overflows container

### PR DESCRIPTION
**Pull request by markocikos**
This is a followup on https://github.com/liferay/liferay-ckeditor/pull/92. See discussion in that PR, as well as under https://issues.liferay.com/browse/LPS-116683.

To test, open editor in message boards. Keep an eye on the `Source` label.

Before:
before

After:
after

Notes:
- I opted to make this change specific to `ClassicEditor` implementation. This includes both JS and CSS. I added the styles to new component scss, that makes sense to me as it is linked to the implementation.
- The solution is mostly skin-independent, although in theory some future skin could override label display and mess up this feature.
- I do not see how this feature could be achieved only with CSS. I also did not find anything useful for it in CKEditor API.
- In the solution, we are comparing toolbars size to container size. If toolbars size is higher, we hide labels.
- You can notice some choppiness when reducing size. This is because we are debouncing resize event.
- As a threshold of hiding, we are not actually taking width of toolbars, but width of toolbars with labels visible (`labeledToolbarsWidth`). If we would take simple current toolbars width, we would end up with toggling behavior on each pixel change, in the area where container size is near toolbars size. The additional size of the label itself would trigger hide, the reduction from removal of label hiding would trigger show.
- I removed  `className="lfr-editable"`. It does nothing, it's not a part of [ckeditor-react API](https://github.com/ckeditor/ckeditor4-react/blob/master/src/ckeditor.jsx). For style purposes, we are using new top level `classic-editor` class.

@wincent @julien @jbalsas @mateomustapic @carloslancha

### CODE REVIEW

**Comment by jbalsas** [comment]
LGTM!

shipit

**Comment by jbalsas** [comment]
Just kidding, not gonna ship it... see if @wincent has feedback on the code, but running tests :)

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js [hunk 1]
```diff
@@ -39,8 +41,10 @@ const ClassicEditor = ({
 	title,
 }) => {
 	const editorRef = useRef();
+	const labeledToolbarsWidth = useRef();
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js:44]: Maybe initial value should be set to 0 here, so `labeledToolbarsWidth.current` will always contain a number, even before `getToolbarsWidth()` is called.

**Comment by jbalsas** [comment]
@markocikos, I was fidgeting around with some other things we need to do for CKEditor and saw that this not working in the default CKEditor version might be an issue if we don't get there in time to replace all versions and maybe even if we do if we ever want to change something.

Could we maybe re-work this as a pure CKEditor plugin? At first I thought this could have a lot of react-ish code, but I think most of react stuff is in there just because and we could do this from a CKEditor plugin in itself using CKEditor imperative APIs.

That would also allow us to bundle it directly inside `liferay-ckeditor` which might make more sense...

What do you think?

**Comment by markocikos** [comment]
Sure, we can do it as a pure CKEditor plugin. I am hoping that `ClassicEditor` becomes the main place for future customization, because it has clear scope, making development easier. CKEditor on the other hand is awkwardly split between repositories, with scripts, odd places for skins and plugins, tmp, diffs, publishing etc. But, I do understand that we still need to support `ckeditor`.

**Comment by jbalsas** [comment]
> Sure, we can do it as a pure CKEditor plugin. I am hoping that ClassicEditor becomes the main place for future customization, because it has clear scope, making development easier. CKEditor on the other hand is awkwardly split between repositories, with scripts, odd places for skins and plugins, tmp, diffs, publishing etc. But, I do understand that we still need to support ckeditor.

Hey @markocikos, there are some valid concerns here. I'm a bit hesitant to bank it all on `ClassicEditor` bypassing the default CKEditor ecosystem. My main concern there is that we'll end up with an inconsistent programming model where it'll be hard to know who does what and where to look for specific features (native plugins, custom react components...)

One thing that could be worth exploring is the release process. I feel like we've improved our CKEditor fork a lot, making the process of patching and adding features more efficient and clear. That being said, since we're moving away from AlloyEditor, the only use for it would be the `liferay-portal` repo, so it might be good to reconsider the approach based on that.

@wincent, @carloslancha, @julien, I think you're the ones that have been more heavily involved in shaping up the build process in the `liferay-ckeditor` repo. Would it be possible, and make sense and would be a good idea moving this directly inside `liferay-portal` to avoid the unnecessary friction?

**Comment by wincent** [comment]
> One thing that could be worth exploring is the release process. I feel like we've improved our CKEditor fork a lot, making the process of patching and adding features more efficient and clear. That being said, since we're moving away from AlloyEditor, the only use for it would be the `liferay-portal` repo, so it might be good to reconsider the approach based on that.
> 
> @wincent, @carloslancha, @julien, I think you're the ones that have been more heavily involved in shaping up the build process in the `liferay-ckeditor` repo. Would it be possible, and make sense and would be a good idea moving this directly inside `liferay-portal` to avoid the unnecessary friction?

- On "more efficient and clear": Definitely more clear, because the patches are now explicitly maintained in the patches directory. Maybe not more efficient because people have to learn the setup/build/patch flow. 🤷‍♂️ What it definitely is, though, is more robust — the old system was basically a confusing mess where it was next to impossible to know how we had diverged from upstream, and with all the manual updates it was all too easy for entropy to creep in and destroy information;  it was kind of like we were running "97% chance that this is version 4.0.11 of CKEditor + random genetic mutations" 🤮 
- On "since we're moving away from AlloyEditor, the only use for it would be in the `liferay-portal` repo": that's actually _already_ the case; AlloyEditor builds using a totally unpatched vanilla version of CKEditor, not the liferay-ckeditor fork. In liferay-portal we use the no-bundled-CKEditor variant of AlloyEditor (instead using the liferay-portal-provided liferay-ckeditor). So, there has only ever been one client of liferay-ckeditor.
- If we move our version of CKEditor into liferay-portal directly and skip the whole release-to-NPM step, then we'll eliminate that part of the friction but potentially lose some of the control that we have in the liferay-ckeditor repo. Specifically, our control of that repo frees us to have a custom build script, and pull in the upstream source via a submodule, as well as store our changes however we want (in our case, we went with patch files). I imagine that if we go to liferay-portal, we'll almost certainly lose the ability to update via a submodule. There are precedents for patch-like files in liferay-portal (eg. `_diffs` files) so we could preserve in some way the pattern of having pristine upstream source (although in this case it would be committed to the repo instead of referenced via a submodule) plus modifications.

There is a bit of a learning curve to doing it the way we do now, but I think there will be in any alternative way that we might do it, so we'd probably want a clear case for the expected benefit that we'd get by switching. I don't know if I see a clear case though. 🤷‍♂️

- ✅ People wouldn't have to be experts on Git submodules any more.
- ✅ No NPM release step.
- ❌ Need to recreate our update/patch/build script (ideally would store it inside liferay-portal, but may need to keep it out if Brian objects to having a weird update process inside the repo).
- ❌ Will still have learning curve, albeit a different one, of a custom update process.

**Comment by jbalsas** [comment]
Okay, just so we're on the same page, @julien is working on turning this into a CKEditor plugin inside `liferay-ckeditor`, so closing here.

**Comment by julien** [comment]
Submitted here [https://github.com/liferay/liferay-ckeditor/pull/102](https://github.com/liferay/liferay-ckeditor/pull/102)

**Pull request closed without merging**